docs/Growing.md

```diff
@@ -415,12 +415,119 @@ with `ctx.redirectTo()`, you can use the flash, accessible through `ctx.flash`.
 ```js
 const flash = await ctx.flash();
 flash.message = 'Everything is fine.';
-return ctx.redirectTo('goodbye');
+await ctx.redirectTo('goodbye');
 ```
 
 Just remember that all session data gets serialized to JSON and stored in encrypted cookies, which usually have a `4096`
 byte (4KiB) limit, depending on browser.
 
+## Final prototype
+
+A final `myapp.js` prototype passing all of the tests above could look like this.
+
+```js
+import mojo from '@mojojs/core';
+import Users from './models/users.js';
+
+// Set cookie secret
+export const app = mojo({secrets: ['Mojolicious rocks']});
+
+app.models.users = new Users();
+
+// Main login action
+app.any('/', async ctx => {
+  // Query parameters
+  const params = await ctx.params();
+  ctx.stash.user = params.get('user')
+  const pass = params.get('pass')
+
+  // Check password and render "index.html.ep" if necessary
+  if (ctx.models.users.check(ctx.stash.user, pass) === false) { 
+    return ctx.render({inline: indexTemplate, inlineLayout: defaultLayout});
+  }
+
+  // Store username in session
+  const session = await ctx.session();
+  session.user = ctx.stash.user;
+
+  // Store a friendly message for the next page in flash
+  const flash = await ctx.flash();
+  flash.message = 'Thanks for logging in.';
+
+  // Redirect to protected page with a 302 response
+  await ctx.redirectTo('protected');
+}); 
+
+// Make sure user is logged in for actions in this action
+const protectedArea = app.under('/protected').to(async ctx => {
+  ctx.stash.session = await ctx.session();
+  if (ctx.stash.session.user !== undefined) return;
+  await ctx.render({text: 'Permission denied'});
+  return false;
+});
+
+// A protected page auto rendering "protected.html.ep"
+protectedArea.get('/').to(async ctx => {
+  ctx.stash.flash = await ctx.flash();
+  return ctx.render({inline: protectedTemplate, inlineLayout: defaultLayout});
+});
+
+// Logout action
+app.get ('/logout', async ctx => {
+  // Expire and in turn clear session automatically
+  const session = await ctx.session();
+  session.expires = 1;
+  // Redirect to main page with a 302 response
+  await ctx.redirectTo('/');
+});
+
+app.start();
+
+const indexTemplate = `
+<form method="post">
+<% if (ctx.stash.user !== null) { %>
+  <b>Wrong name or password, please try again.</b><br>
+<% } %>
+  User:<br>
+  <input name="user">
+  <br>Password:<br>
+  <input type="password" name="pass">
+  <br>
+<%= ctx.tag('button', { type: 'Submit'}, 'Login') %>
+</form>
+`;
+
+const protectedTemplate = `
+% if (ctx.stash.flash.message != null) {
```

Comment:
Writing out `ctx.stash.flash.message` kinda defeats the purpose of assigning anything to the stash in the first place.

Reply:
Simply adding a `% const flash = await ctx.flash();` to the top would be cleaner.

Reply:
Templates are always `async`, so you can use `await` freely.